ui/hooks: page actions

kanish671 opened the pull request:
- [x] Identified the issue which this PR solves.
- [x] Read the [**CONTRIBUTING**](https://github.com/target/goalert/blob/master/CONTRIBUTING.md) document.
- [x] Code builds clean without any errors or warnings.
- [x] Added appropriate tests for any new functionality.
- [x] All new and existing tests passed.
- [x] Added comments in the code, where necessary.
- [x] Ran `make check` to catch common errors. Fixed any that came up.

**Description:**
Converted PageActions to use hooks

**Which issue(s) this PR fixes:**
Part of #923

Forfold commented on web/src/app/util/PageActions.js:
```diff
@@ -8,104 +8,90 @@ const PageActionsContext = React.createContext({
-    this._pending = false
-    this.setActions.cancel()
-  }
+  let _mountCount = 0
+  let _mounted = false
+  let _pending = null
```
I think the way these are evaluated in classes vs. hooks is a bit different than it seems- `this.foo = 'bar'` is a synchronous update, and that value will live in the "background" of the component allowing it to retain its value on every re-render. With hooks though, these will get reset back to `0`, `false`, and `null` on every re-render.

I think we can get around this in hooks if we initialize the variables outside of the component and use the `useEffect` hook, as it will only update the values on mount and unmount:

```js
let _mountCount = 0
let _mounted = false

const MyComponentExample = () => {
  useEffect(() => {
    // componentDidMount code here

    return () => {
      // componentWillUnmount code here
    }
  }, [])

  // ...
}
```

mastercactapus replied:
Having them outside of the component could introduce really odd side-effects bugs/leaks, we can use `useRef` though -- https://reactjs.org/docs/hooks-faq.html#is-there-something-like-instance-variables

And then `useEffect(..., [])` to handle the mount/unmount triggers.

kanish671 replied:
Hey thanks! I'll take a look at it and make the changes

kanish671 replied:
Sure, I'll take a look at it.

kanish671 replied:
I've made some changes. please take a look and let me know if it looks good.

mastercactapus commented on web/src/app/util/PageActions.js:
```diff
@@ -8,104 +8,91 @@ const PageActionsContext = React.createContext({
+      _pending.current = false
+      debouncedSetActions.cancel()
+    }
+  })
```
Adding an empty dependency array at the end will ensure this `useEffect` only runs once on mount and once on unmount -- otherwise, it's going to call `debouncedSetActions` and `.cancel` on every render.
```suggestion
  }, [])
```

mastercactapus commented on web/src/app/util/PageActions.js:
```diff
@@ -8,104 +8,91 @@ const PageActionsContext = React.createContext({
+      props.trackMount(false)
+      props.setActions(null)
     }
+  })
```
Same as above, just add an empty array to have it only run on mount/unmount.

```suggestion
  }, [])
```

kanish671 commented:
@mastercactapus Addressed the review comments. Also, I had missed out a piece of code that was causing bugs when I was testing locally. Added that too.

kanish671 commented:
@mastercactapus @Forfold is there anything else that needs to be done here?

Forfold commented on web/src/app/util/PageActions.js:
```diff
@@ -8,104 +8,95 @@ const PageActionsContext = React.createContext({
-      </PageActionsContext.Provider>
-    )
-  }
+  useEffect(() => {
+    _mounted.current = true
+    if (_pending.current) {
+      debouncedSetActions(_pending.current)
+      _pending.current = false
+    }
+    return () => {
+      _mounted.current = false
+      _pending.current = false
+      debouncedSetActions.cancel()
+    }
+  }, [])
```
nit: move this useEffect block to the top of the component, under the `const _pending = useRef(null)` line

kanish671 replied:
I can move it under `const debouncedSetActions = debounce(_setActions)`. Otherwise,  eslint complains about using `debouncedSetActions` before defining it.

Forfold reviewed (approved):
LGTM! Only have 1 comment regarding formatting, so approving for CI + smoketests

kanish671 commented:
@mastercactapus could you also review this again?

kanish671 commented:
Thank you! 😄
